File: src/main/java/com/ghostchu/peerbanhelper/BuildMeta.java
package com.ghostchu.peerbanhelper;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.Properties;

@Data
@NoArgsConstructor
@AllArgsConstructor
public final class BuildMeta {
    private String version = "999.999.999-undefined";
    private String os = "Unknown";
    private String branch = "Unknown";
    private String commit = "Unknown";
    private String abbrev = "Unknown";
    private String compileTime = "Unknown";
    private String compileHost;
    private String compileUser;
    private String compileEmail;

    public void loadBuildMeta(Properties configuration) {
        this.version = ExternalSwitch.parse("pbh.buildmeta.maven.version", configuration.getProperty("git.build.version", "0.0.0-undefined"));
        this.branch = ExternalSwitch.parse("pbh.buildmeta.git.branch", configuration.getProperty("git.branch", "Unknown"));
        this.commit = configuration.getProperty("git.commit.id", "Unknown");
        this.abbrev = configuration.getProperty("git.commit.id.abbrev", "Unknown");
        this.os = System.getProperty("os.name", "Unknown");
        this.compileTime = configuration.getProperty("git.commit.time", "Unknown");
        this.compileHost = configuration.getProperty("git.build.host", "Unknown");
        this.compileUser = configuration.getProperty("git.commit.user.name", "Unknown");
        this.compileEmail = configuration.getProperty("git.commit.user.email", "Unknown");
    }

    public boolean isSnapshotOrBeta() {
        return !"release".equalsIgnoreCase(branch);
    }

    public String toString() {
        return "BuildMeta(version=" + this.getVersion() + ")";
    }
}

File: src/main/java/com/ghostchu/peerbanhelper/BuildMeta.kt
package com.ghostchu.peerbanhelper

import org.bspfsystems.yamlconfiguration.file.YamlConfiguration

open class BuildMeta {
    var version: String = "unknown"
    private var isNativeImage: Boolean = false

    fun loadBuildMeta(configuration: YamlConfiguration) {
        this.version = configuration.getString("maven.version")!!
        this.isNativeImage = configuration.getString("env.native-image", "false").equals("true", ignoreCase = true)
    }

    override fun equals(o: Any?): Boolean {
        if (o === this) return true
        if (o !is BuildMeta) return false
        if (!o.canEqual(this as Any)) return false
        val `this$version`: Any = this.version
        val `other$version`: Any = o.version
        if (`this$version` != `other$version`) return false
        return true
    }

    protected fun canEqual(other: Any?): Boolean {
        return other is BuildMeta
    }

    override fun hashCode(): Int {
        val PRIME = 59
        val result = 1
        val version = this.version
        return result * PRIME + version.hashCode()
    }

    override fun toString(): String {
        return "BuildMeta(version=" + this.version + ")"
    }
}
